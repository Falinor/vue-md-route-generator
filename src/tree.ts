export type Tree<T> = { value: T; children?: Tree<T>[] }

type ForEachFunction<T> = (node: T) => void

export function forEach<T = any>(tree: Tree<T>, fn: ForEachFunction<T>): void {
  const { value, children } = tree
  fn(value)

  if (!children) {
    return
  }

  children.forEach(child => {
    forEach(child, fn)
  })
}

type MapFunction<T, U> = (value: T, children?: Tree<T>[]) => U

export function map<T = any, U = any>(
  tree: Tree<T>,
  fn: MapFunction<T, U>
): Tree<U> {
  const { value, children } = tree

  if (!children) {
    return { value: fn(value) }
  }

  return {
    value: fn(value, children),
    children: children.map(child => map(child, fn))
  }
}

export function flatten<T>(tree: Tree<T>): T[] {
  function flattenRec<T>(acc: T[], node: Tree<T>): T[] {
    acc.push(node.value)
    if (node.children) {
      acc.concat(...node.children.map(child => flattenRec(acc, child)))
    }
    return acc
  }
  return flattenRec<T>([], tree)
}

export function flatMap<T = any, U = any>(
  tree: Tree<T>,
  fn: MapFunction<T, U>
): U[] {
  return flatten(map(tree, fn))
}

export type SortFunction<T> = (a: Tree<T>, b: Tree<T>) => number

export function sort<T = any>(tree: Tree<T>, fn: SortFunction<T>): Tree<T> {
  const { value, children } = tree

  if (!children) {
    return tree
  }

  return {
    value,
    children: [...children].map(child => sort(child, fn)).sort(fn)
  }
}

export type FilterFunction<T> = (value: T, children?: Tree<T>[]) => boolean

export function filter<T = any>(
  tree: Tree<T>,
  fn: FilterFunction<T>
): Tree<T> | null {
  if (!fn(tree.value)) {
    return null
  }

  if (!tree.children) {
    return tree
  }

  return {
    value: tree.value,
    children: tree.children
      .map(child => filter(child, fn))
      .filter(v => v !== null) as Tree<T>[]
  }
}
