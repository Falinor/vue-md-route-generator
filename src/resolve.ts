import parseFrontMatter from 'gray-matter'
import isEmpty from 'lodash.isempty'
import path from 'path'

import { filter, map, sort, Tree } from './tree'

export interface PageMeta {
  name: string
  specifier: string
  path: string
  pathSegments: string[]
  component: string | null
  routeMeta?: unknown
}

export interface File {
  path: string
  type: 'directory' | 'file'
}

export type FileTree = Tree<File>
export type PageMetaTree = Tree<PageMeta>

export function resolveRoutePaths(
  tree: FileTree,
  importPrefix: string,
  readFile: (path: string) => string
): PageMetaTree {
  return sortByMeta(
    filterDirectoryMeta(
      map(tree, (file: File, children?: FileTree[]) => {
        const segments: string[] = file.path.split('/')
        const meta: PageMeta = {
          name: pathToName(segments),
          path: pathToRoute(segments, 0, false),
          pathSegments: toActualPath(segments),
          component: null,
          specifier: pathToSpecifier(segments)
        }

        if (file.type === 'file') {
          const content = readFile(file.path)
          const { data } = parseFrontMatter(content)
          if (!isEmpty(data)) {
            meta.routeMeta = data
          }
          meta.component = importPrefix + file.path
        }

        if (file.type === 'directory' && children?.length) {
          const metaFile = children.find(
            child =>
              child.value.type === 'file' &&
              basename(child.value.path) === '__meta__'
          )
          if (metaFile !== undefined) {
            const content = readFile(metaFile.value.path)
            const { data } = parseFrontMatter(content)
            if (!isEmpty(data)) {
              meta.routeMeta = data
            }
          }
        }

        return meta
      })
    )
  )
}

function sortByMeta(tree: PageMetaTree): PageMetaTree {
  return sort(tree, (a: PageMetaTree, b: PageMetaTree): number => {
    const routeMeta1 = a.value.routeMeta
    const routeMeta2 = b.value.routeMeta

    if (hasOrder(routeMeta1)) {
      if (hasOrder(routeMeta2)) {
        return routeMeta1.order - routeMeta2.order
      }
    }

    if (hasOrder(routeMeta2)) {
      return 1
    }

    return 0
  })
}

function hasOrder(obj: unknown): obj is { order: number } {
  return typeof obj === 'object' && obj !== null && obj.hasOwnProperty('order')
}

function filterDirectoryMeta(tree: PageMetaTree): PageMetaTree {
  return filter(
    tree,
    meta => !meta.component?.endsWith('__meta__.md')
  ) as PageMetaTree
}

function isOmittable(segment: string): boolean {
  return segment === 'index'
}

/**
 * - Remove `.md` from the last path
 * - Omit if the last segment is `index`
 */
function toActualPath(segments: string[]): string[] {
  const lastIndex = segments.length - 1
  const last = basename(segments[lastIndex])

  const segmentsWithoutLast = segments.slice(0, -1)
  return isOmittable(last)
    ? segmentsWithoutLast
    : segmentsWithoutLast.concat(last)
}

function pathToName(segments: string[]): string {
  const last = segments[segments.length - 1]
  return segments
    .slice(0, -1)
    .concat(basename(last))
    .join('-')
}

function pathToSpecifier(segments: string[]): string {
  const name = pathToName(segments)
  const replaced = name
    .replace(/(^|[^a-zA-Z])([a-zA-Z])/g, (_, a, b) => {
      return a + b.toUpperCase()
    })
    .replace(/[^a-zA-Z0-9]/g, '')

  return /^\d/.test(replaced) ? '_' + replaced : replaced
}

function pathToRoute(
  segments: string[],
  parentDepth: number,
  nested: boolean
): string {
  const prefix = nested || parentDepth > 0 ? '' : '/'
  return (
    prefix +
    toActualPath(segments)
      .slice(parentDepth)
      .join('/')
  )
}

function basename(filename: string): string {
  return path.basename(filename, '.md')
}
