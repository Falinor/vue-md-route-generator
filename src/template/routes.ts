import { PageMeta } from '../resolve'
import { flatMap, Tree } from '../tree'

export type PageMetaTree = Tree<PageMeta>

function createChildrenRoute(children: PageMetaTree[]): string {
  return `children: [${children.map(createRoute).join(',')}],`
}

function createRoute(meta: PageMetaTree): string {
  const children = !meta.children ? '' : createChildrenRoute(meta.children)

  // If default child exists, the route should not have a name.
  const routeName = meta.children?.some(m => m.value.path === '')
    ? ''
    : `name: '${meta.value.name}',`

  const routeMeta = meta.value.routeMeta
    ? `meta: ${JSON.stringify(meta.value.routeMeta, null, 2)},`
    : ''

  const redirect: PageMetaTree =
    meta.children?.find(child => child.value.specifier.endsWith('Index')) ??
    (meta.children?.[0] as PageMetaTree)
  const specifier = meta.value.component
    ? `component: ${meta.value.specifier},`
    : `redirect: { name: '${redirect.value.name}' },`

  return `
  {
    ${routeName}
    path: '${meta.value.path}',
    ${specifier}${routeMeta}${children}
  }
  `
}

function createImport(
  meta: PageMeta,
  dynamic: boolean,
  chunkNamePrefix: string
): string {
  if (meta.component === null) {
    return ''
  }

  return dynamic
    ? `function ${meta.specifier}() { return import(/* webpackChunkName: "${chunkNamePrefix}${meta.name}" */ '${meta.component}') }`
    : `import ${meta.specifier} from '${meta.component}'`
}

export interface RouteString {
  imports: string
  code: string
}

export function createRoutes(
  meta: PageMetaTree,
  dynamic: boolean,
  chunkNamePrefix: string
): RouteString {
  const imports: string = flatMap(meta, node =>
    createImport(node, dynamic, chunkNamePrefix)
  ).join('\n')

  const code = createRoute(meta)
  return { imports, code }
}
