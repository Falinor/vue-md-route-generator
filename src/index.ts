import directoryTree, { DirectoryTree } from 'directory-tree'
import fs from 'fs'
import { basename, dirname, isAbsolute, join } from 'path'
import prettier from 'prettier'

import { FileTree, resolveRoutePaths } from './resolve'
import { createRoutes, RouteString } from './template/routes'

export interface GenerateConfig {
  folders: string[]
  importPrefix?: string
  dynamicImport?: boolean
  chunkNamePrefix?: string
  nested?: boolean
}

function relativeTo(prefix: string, path: string): string {
  const segments = path.split('/')
  const index = segments.findIndex(seg => seg === prefix)
  return segments.slice(index).join('/')
}

function toPathTree(directoryTree: DirectoryTree, prefix: string): FileTree {
  return {
    value: {
      path: isAbsolute(directoryTree.path)
        ? relativeTo(prefix, directoryTree.path)
        : directoryTree.path,
      type: directoryTree.type
    },
    children: directoryTree?.children?.map(child => toPathTree(child, prefix))
  }
}

export function generateRoutes({
  folders,
  importPrefix = '@/assets/',
  dynamicImport = true,
  chunkNamePrefix = ''
}: GenerateConfig): string {
  const routeStrings: RouteString[] = folders.map(folder => {
    const dirTree = directoryTree(folder, {
      extensions: /\.md$/,
      normalizePath: true
    })
    const pathTree = toPathTree(dirTree, basename(folder))

    const meta = resolveRoutePaths(pathTree, importPrefix, file => {
      return fs.readFileSync(join(dirname(folder), file), 'utf8')
    })
    return createRoutes(meta, dynamicImport, chunkNamePrefix)
  })
  const rendererImport: string = `
  import Vue from 'vue'
  
  Vue.component('renderer', resolve => {
    setTimeout(() => resolve({
      render: createElement => createElement('router-view')
    }))
  })
  `
  const imports: string = [rendererImport]
    .concat(routeStrings.map(rs => rs.imports))
    .join('\n')
  const routes: string = routeStrings.map(rs => rs.code).join(',\n')
  return prettier.format(`${imports}\n\nexport default [${routes}]`, {
    parser: 'babel',
    semi: false,
    singleQuote: true
  })
}
