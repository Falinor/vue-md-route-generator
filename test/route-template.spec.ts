import { createRoutes } from '../src/template/routes'
import { PageMeta } from '../src/resolve'

describe('Route template', () => {
  it('should generate routes', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/assets/foo.md'
      },
      {
        name: 'bar',
        specifier: 'Bar',
        path: '/bar',
        pathSegments: ['bar'],
        component: '@/assets/bar.md'
      }
    ]

    expect(createRoutes(meta, true, '')).toMatchSnapshot()
  })

  it('should generate nested routes', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/assets/foo.md',
        children: [
          {
            name: 'bar',
            specifier: 'FooBar',
            path: 'bar',
            pathSegments: ['foo', 'bar'],
            component: '@/assets/bar.md'
          },
          {
            name: 'baz',
            specifier: 'FooBaz',
            path: 'baz',
            pathSegments: ['foo', 'baz'],
            component: '@/assets/baz.md'
          }
        ]
      }
    ]

    expect(createRoutes(meta, true, '')).toMatchSnapshot()
  })

  it('should generate static import code', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/assets/foo.md'
      },
      {
        name: 'bar',
        specifier: 'Bar',
        path: '/bar',
        pathSegments: ['bar'],
        component: '@/assets/bar.md'
      }
    ]

    expect(createRoutes(meta, false, '')).toMatchSnapshot()
  })

  it('should generate route meta', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/assets/foo.md',
        routeMeta: {
          title: 'Hello'
        }
      }
    ]

    expect(createRoutes(meta, false, '')).toMatchSnapshot()
  })

  it('should configure chunk name prefix', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/assets/foo.md'
      },
      {
        name: 'bar',
        specifier: 'Bar',
        path: '/bar',
        pathSegments: ['bar'],
        component: '@/assets/bar.md'
      }
    ]

    expect(createRoutes(meta, true, 'page-')).toMatchSnapshot()
  })

  it('should not include name if the route has a default child', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        specifier: 'Foo',
        path: '/foo',
        pathSegments: ['foo'],
        component: '@/assets/foo.md',
        children: [
          {
            name: 'foo-index',
            specifier: 'FooIndex',
            path: '',
            pathSegments: ['foo'],
            component: '@/pages/foo/index.md'
          }
        ]
      }
    ]

    expect(createRoutes(meta, true, '')).toMatchSnapshot()
  })
})
