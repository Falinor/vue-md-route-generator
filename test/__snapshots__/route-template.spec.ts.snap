// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route template should configure chunk name prefix 1`] = `
"function Foo() {
  return import(/* webpackChunkName: \\"page-foo\\" */ '@/assets/foo.md')
}
function Bar() {
  return import(/* webpackChunkName: \\"page-bar\\" */ '@/assets/bar.md')
}

export default [
  {
    name: 'foo',
    path: '/foo',
    component: Foo
  },
  {
    name: 'bar',
    path: '/bar',
    component: Bar
  }
]
"
`;

exports[`Route template should generate nested routes 1`] = `
"function Foo() {
  return import(/* webpackChunkName: \\"foo\\" */ '@/assets/foo.md')
}
function FooBar() {
  return import(/* webpackChunkName: \\"bar\\" */ '@/assets/bar.md')
}
function FooBaz() {
  return import(/* webpackChunkName: \\"baz\\" */ '@/assets/baz.md')
}

export default [
  {
    name: 'foo',
    path: '/foo',
    component: Foo,
    children: [
      {
        name: 'bar',
        path: 'bar',
        component: FooBar
      },
      {
        name: 'baz',
        path: 'baz',
        component: FooBaz
      }
    ]
  }
]
"
`;

exports[`Route template should generate route meta 1`] = `
"import Foo from '@/assets/foo.md'

export default [
  {
    name: 'foo',
    path: '/foo',
    component: Foo,
    meta: {
      title: 'Hello'
    }
  }
]
"
`;

exports[`Route template should generate routes 1`] = `
"function Foo() {
  return import(/* webpackChunkName: \\"foo\\" */ '@/assets/foo.md')
}
function Bar() {
  return import(/* webpackChunkName: \\"bar\\" */ '@/assets/bar.md')
}

export default [
  {
    name: 'foo',
    path: '/foo',
    component: Foo
  },
  {
    name: 'bar',
    path: '/bar',
    component: Bar
  }
]
"
`;

exports[`Route template should generate static import code 1`] = `
"import Foo from '@/assets/foo.md'
import Bar from '@/assets/bar.md'

export default [
  {
    name: 'foo',
    path: '/foo',
    component: Foo
  },
  {
    name: 'bar',
    path: '/bar',
    component: Bar
  }
]
"
`;

exports[`Route template should not include name if the route has a default child 1`] = `
"function Foo() {
  return import(/* webpackChunkName: \\"foo\\" */ '@/assets/foo.md')
}
function FooIndex() {
  return import(/* webpackChunkName: \\"foo-index\\" */ '@/pages/foo/index.md')
}

export default [
  {
    path: '/foo',
    component: Foo,
    children: [
      {
        name: 'foo-index',
        path: '',
        component: FooIndex
      }
    ]
  }
]
"
`;
